File: frontend/evarkadasim-yeni-main/services/propertyService.ts
import { Property, PropertyMapPin } from '@/types';
import { apiClient } from './apiClient';

interface PropertyDto {
  id: number;
  title: string;
  price: string;
  priceAmount: number;
  currency: string;
  pricePeriod: string;
  location: string;
  bedrooms: number;
  bathrooms: number;
  images: string[];
  description?: string;
  amenities: string[];
  availableFrom: string;
  // JsonStringEnumConverter: "Apartment" | "Studio" | "House" | "Room"
  propertyType: 'Apartment' | 'Studio' | 'House' | 'Room';
  furnished: boolean;
  petsAllowed: boolean;
  smokingAllowed: boolean;
  latitude?: number;
  longitude?: number;
  ownerId: string;
  ownerName: string;
}

interface PropertyMapPinDto {
  id: number;
  latitude: number;
  longitude: number;
  title: string;
  price: string;
  location: string;
  propertyType: 'Apartment' | 'Studio' | 'House' | 'Room';
  ownerId: string;
  ownerName: string;
}

// Backend "Room" (oda kiralama) → frontend "shared" (en yakın karşılık)
function mapPropertyType(val: string): Property['propertyType'] {
  const map: Record<string, Property['propertyType']> = {
    Apartment: 'apartment',
    Studio: 'studio',
    House: 'house',
    Room: 'shared',
  };
  return map[val] ?? 'apartment';
}

function toProperty(dto: PropertyDto): Property {
  return {
    id: dto.id.toString(),
    title: dto.title,
    price: dto.price,
    location: dto.location,
    bedrooms: dto.bedrooms,
    bathrooms: dto.bathrooms,
    images: dto.images,
    description: dto.description ?? '',
    amenities: dto.amenities,
    availableFrom: new Date(dto.availableFrom),
    propertyType: mapPropertyType(dto.propertyType),
    furnished: dto.furnished,
    petsAllowed: dto.petsAllowed,
    smokingAllowed: dto.smokingAllowed,
    latitude: dto.latitude,
    longitude: dto.longitude,
    ownerId: dto.ownerId,
    ownerName: dto.ownerName,
  };
}

function toMapPin(dto: PropertyMapPinDto): PropertyMapPin {
  return {
    id: dto.id.toString(),
    latitude: dto.latitude,
    longitude: dto.longitude,
    title: dto.title,
    price: dto.price,
    location: dto.location,
    propertyType: mapPropertyType(dto.propertyType),
    ownerId: dto.ownerId,
    ownerName: dto.ownerName,
  };
}

export interface PropertyFormData {
  title: string;
  location: string;
  priceAmount: number;
  currency: string;
  pricePeriod: string;
  bedrooms: number;
  bathrooms: number;
  propertyType: 'Apartment' | 'Studio' | 'House' | 'Room';
  furnished: boolean;
  petsAllowed: boolean;
  smokingAllowed: boolean;
  description: string;
  availableFrom: string;
  images?: string[];
  latitude?: number;
  longitude?: number;
}

export { PropertyDto };

export const propertyService = {
  getList: async (skip = 0, take = 20): Promise<Property[]> => {
    const { data } = await apiClient.get<PropertyDto[]>('/property', { params: { skip, take } });
    return data.map(toProperty);
  },

  getMapPins: async (city?: string): Promise<PropertyMapPin[]> => {
    const { data } = await apiClient.get<PropertyMapPinDto[]>('/property/map', {
      params: city ? { city } : undefined,
    });
    return data.map(toMapPin);
  },

  getMine: async (): Promise<PropertyDto | null> => {
    const { data, status } = await apiClient.get<PropertyDto>('/property/mine', {
      validateStatus: (s) => s === 200 || s === 204,
    });
    return status === 204 ? null : data;
  },

  create: async (form: PropertyFormData): Promise<PropertyDto> => {
    const { data } = await apiClient.post<PropertyDto>('/property', {
      title: form.title,
      priceAmount: form.priceAmount,
      currency: form.currency,
      pricePeriod: form.pricePeriod,
      location: form.location,
      bedrooms: form.bedrooms,
      bathrooms: form.bathrooms,
      propertyType: form.propertyType,
      furnished: form.furnished,
      petsAllowed: form.petsAllowed,
      smokingAllowed: form.smokingAllowed,
      description: form.description,
      availableFrom: form.availableFrom,
      images: form.images ?? [],
      amenities: [],
      latitude: form.latitude,
      longitude: form.longitude,
    });
    return data;
  },

  update: async (id: number, form: PropertyFormData): Promise<PropertyDto> => {
    const { data } = await apiClient.put<PropertyDto>(`/property/${id}`, {
      title: form.title,
      priceAmount: form.priceAmount,
      currency: form.currency,
      pricePeriod: form.pricePeriod,
      location: form.location,
      bedrooms: form.bedrooms,
      bathrooms: form.bathrooms,
      propertyType: form.propertyType,
      furnished: form.furnished,
      petsAllowed: form.petsAllowed,
      smokingAllowed: form.smokingAllowed,
      description: form.description,
      availableFrom: form.availableFrom,
      images: form.images ?? [],
      amenities: [],
      latitude: form.latitude,
      longitude: form.longitude,
    });
    return data;
  },

  delete: async (id: number): Promise<void> => {
    await apiClient.delete(`/property/${id}`);
  },

  deleteMine: async (): Promise<void> => {
    await apiClient.delete('/property/mine');
  },
};

File: backend/EvArkadasimV2.Application/Services/PropertyService.cs
using EvArkadasimV2.Application.DTOs.Property;
using EvArkadasimV2.Application.Exceptions;
using EvArkadasimV2.Application.Interfaces.Repositories;
using EvArkadasimV2.Application.Interfaces.Services;
using EvArkadasimV2.Domain.Entities;
using EvArkadasimV2.Domain.Enums;

namespace EvArkadasimV2.Application.Services
{
    public class PropertyService : IPropertyService
    {
        private const int DefaultTake = 20;
        private const int MaxTake = 50;

        private readonly IPropertyRepository _propertyRepository;
        private readonly IUserRepository _userRepository;

        public PropertyService(IPropertyRepository propertyRepository, IUserRepository userRepository)
        {
            _propertyRepository = propertyRepository;
            _userRepository = userRepository;
        }

        public async Task<IEnumerable<PropertyDto>> GetListAsync(
            string? location,
            PropertyType? propertyType,
            decimal? maxPrice,
            int? bedrooms,
            bool? petsAllowed,
            int skip,
            int take)
        {
            if (skip < 0) skip = 0;
            if (take <= 0) take = DefaultTake;
            if (take > MaxTake) take = MaxTake;

            var properties = await _propertyRepository.GetFilteredAsync(
                location, propertyType, maxPrice, bedrooms, petsAllowed, skip, take);

            return properties.Select(MapToDto);
        }

        public async Task<PropertyDto> GetByIdAsync(int id)
        {
            var property = await _propertyRepository.GetByIdWithOwnerAsync(id);
            if (property == null)
                throw new NotFoundException($"İlan bulunamadı. Id: {id}");

            return MapToDto(property);
        }

        public async Task<PropertyDto> CreateAsync(string ownerId, CreatePropertyDto dto)
        {
            var property = new Property
            {
                Title = dto.Title,
                PriceAmount = dto.PriceAmount,
                Currency = dto.Currency,
                PricePeriod = dto.PricePeriod,
                Location = dto.Location,
                Bedrooms = dto.Bedrooms,
                Bathrooms = dto.Bathrooms,
                Images = dto.Images,
                Description = dto.Description ?? string.Empty,
                Amenities = dto.Amenities,
                AvailableFrom = dto.AvailableFrom,
                PropertyType = dto.PropertyType,
                Furnished = dto.Furnished,
                PetsAllowed = dto.PetsAllowed,
                SmokingAllowed = dto.SmokingAllowed,
                OwnerId = ownerId,
                Latitude = dto.Latitude,
                Longitude = dto.Longitude
            };

            await _propertyRepository.AddAsync(property);
            await _propertyRepository.SaveChangesAsync();

            var owner = await _userRepository.GetUserWithProfileAsync(ownerId);
            if (owner?.Profile != null && owner.Profile.LookingFor != LookingFor.Roommate)
            {
                owner.Profile.LookingFor = LookingFor.Roommate;
                await _userRepository.SaveChangesAsync();
            }

            // Owner navigation property olmadan dönmemek için tekrar çek
            var created = await _propertyRepository.GetByIdWithOwnerAsync(property.Id);
            return MapToDto(created!);
        }

        public async Task<PropertyDto> UpdateAsync(int id, string currentUserId, UpdatePropertyDto dto)
        {
            var property = await _propertyRepository.GetByIdAsync(id);
            if (property == null)
                throw new NotFoundException($"İlan bulunamadı. Id: {id}");

            if (property.OwnerId != currentUserId)
                throw new ForbiddenException("Bu ilanı düzenleme yetkiniz yok.");

            property.Title = dto.Title;
            property.PriceAmount = dto.PriceAmount;
            property.Currency = dto.Currency;
            property.PricePeriod = dto.PricePeriod;
            property.Location = dto.Location;
            property.Bedrooms = dto.Bedrooms;
            property.Bathrooms = dto.Bathrooms;
            property.Images = dto.Images;
            property.Description = dto.Description ?? string.Empty;
            property.Amenities = dto.Amenities;
            property.AvailableFrom = dto.AvailableFrom;
            property.PropertyType = dto.PropertyType;
            property.Furnished = dto.Furnished;
            property.PetsAllowed = dto.PetsAllowed;
            property.SmokingAllowed = dto.SmokingAllowed;
            property.Latitude = dto.Latitude;
            property.Longitude = dto.Longitude;

            _propertyRepository.Update(property);
            await _propertyRepository.SaveChangesAsync();

            var updated = await _propertyRepository.GetByIdWithOwnerAsync(id);
            return MapToDto(updated!);
        }

        public async Task DeleteAsync(int id, string currentUserId)
        {
            var property = await _propertyRepository.GetByIdAsync(id);
            if (property == null)
                throw new NotFoundException($"İlan bulunamadı. Id: {id}");

            if (property.OwnerId != currentUserId)
                throw new ForbiddenException("Bu ilanı silme yetkiniz yok.");

            _propertyRepository.Remove(property);
            await _propertyRepository.SaveChangesAsync();
        }

        public async Task<IEnumerable<PropertyMapPinDto>> GetMapPinsAsync(string? city = null)
        {
            var properties = await _propertyRepository.GetWithCoordinatesAsync();

            if (!string.IsNullOrEmpty(city))
                properties = properties.Where(p =>
                    p.Location.Contains(city, StringComparison.OrdinalIgnoreCase)).ToList();

            return properties.Select(p => new PropertyMapPinDto
            {
                Id = p.Id,
                Latitude = p.Latitude!.Value,
                Longitude = p.Longitude!.Value,
                Title = p.Title,
                Price = $"{p.Currency}{p.PriceAmount:N0}/{p.PricePeriod}",
                Location = p.Location,
                PropertyType = p.PropertyType,
                OwnerId = p.OwnerId,
                OwnerName = $"{p.Owner?.Name} {p.Owner?.LastName}".Trim()
            });
        }

        private static PropertyDto MapToDto(Property p) => new()
        {
            Id = p.Id,
            Title = p.Title,
            Price = $"{p.Currency}{p.PriceAmount:N0}/{p.PricePeriod}",
            Location = p.Location,
            Bedrooms = p.Bedrooms,
            Bathrooms = p.Bathrooms,
            Images = p.Images ?? new List<string>(),
            Description = p.Description,
            Amenities = p.Amenities ?? new List<string>(),
            AvailableFrom = p.AvailableFrom,
            PropertyType = p.PropertyType,
            Furnished = p.Furnished,
            PetsAllowed = p.PetsAllowed,
            SmokingAllowed = p.SmokingAllowed,
            Latitude = p.Latitude,
            Longitude = p.Longitude,
            OwnerId = p.OwnerId,
            OwnerName = p.Owner?.Name ?? string.Empty
        };
    }
}
